src/lib/reasoners/intervals_core.ml

```diff
@@ -0,0 +1,811 @@
+      | None, Some _ -> true
+      | Some (i1, s1), Some (i2, s2) ->
+        subset_seq ~strict i1 s1 i2 s2
+
+    let equal u1 u2 =
+      Seq.equal Interval.equal (to_seq u1) (to_seq u2)
+
+    let checked ((glb, u', gub) as u) =
+      let rec loop ex = function
+        | [] -> Empty ex
+        | Empty ex' :: u' -> loop (Ex.union ex ex') u'
+        | NonEmpty _ :: _ -> NonEmpty u
+      in
+      if gub < glb then Empty Ex.empty else loop Ex.empty u'
+
+    let nonempty = function
+      | NonEmpty u -> u
+      | Empty _ -> invalid_arg "nonempty"
+
+    let checked_opt = function
+      | None -> Empty Ex.empty
+      | Some u -> checked u
+
+    let add_explanation ~ex ((_, u, _) as u') =
+      if Ex.is_empty ex then u' else
+        let[@tail_mod_cons] rec loop ex' = function
+          | [] -> [ Empty ex' ]
+          | [ NonEmpty i ] as u when i.ub = OT.pinfty -> Empty ex' :: u
+          | NonEmpty i :: u -> Empty ex' :: NonEmpty i :: loop ex u
+          | Empty ex'' :: u -> loop (Ex.union ex' ex'') u
+        in
+        let u =
+          match u with
+          | NonEmpty i :: u when i.lb = OT.minfty ->
+            if i.ub = OT.pinfty then [ NonEmpty i ] else NonEmpty i :: loop ex u
+          | _ -> loop ex u
+        in
+        (OT.minfty, u, OT.pinfty)
+
+    let min_ex ex1 ex2 =
+      if Stdlib.(Ex.compare ex1 ex2 <= 0) then ex1 else ex2
+
+    (* Computing the intersection follows a fairly natural, if a bit tedious,
+       process: we are simply iterating on the two unions at the same time and
+       computing the intersections as we go.
+
+       Some care need to be taken of to make sure to use appropriate
+       explanations. When one of the unions is enough to forbid a given
+       forbidden interval, we only want to use that one (for instance, the
+       intersection of [-oo; 4] with explanation [ex1] and [-oo; 10] with
+       explanation [ex2] only needs explanation [ex1]). We only need to take the
+       union of the explanations when the explanation in one union is used to
+       skip over an interval that was allowed in the other union; for instance,
+       the intersection of
+
+        [0; 1] U !ex1 U [2; 3]
+
+       and
+
+        !ex2 U [4; 10]
+
+       must be
+
+        !(union ex1 ex2) U [2; 3]
+
+       We need [ex1] to forbid ]1; 2[ and [ex2] to forbid [0; 1] and [2; 4].
+
+       In addition, the current implementation respects the known global
+       lower/upper bounds, which adds a low amount of complexity (everything
+       before/after the global lower/upper bound can be removed without
+       justification). An alternative implementation could ignore the global
+       lower and upper bounds and always return [-oo, +oo] as an enclosing
+       interval; which might improve efficiency of further computations. *)
+    let intersect =
+      let[@tail_mod_cons] rec loop ex1 u1 ex2 u2 gub =
+        match u1, u2 with
+        | [], [] -> (empty [@tailcall false]) (min_ex ex1 ex2) []
```

Comment:
Why do you use this annotation here? Is it crucial to ensure that the tail call is optimized?